### PR TITLE
[Reviewer: Tom] Reject on queue

### PR DESCRIPTION
**Pull request by eleanor-merry**
This PR ages out old requests if they've been on the queue for too long; rejecting them in the same way that we reject on overload.

It's not been tested, and I've not thought about what a sensible default would be.

### CODE REVIEW

File: include/cfgoptions.h [hunk 1]
```diff
@@ -147,6 +147,7 @@ struct options
   bool                                 reject_if_no_matching_ifcs;
   std::string                          dummy_app_server;
   bool                                 http_acr_logging;
+  int                                  request_on_queue_timeout;
```
Comment (tomwatson1024) [inline include/cfgoptions.h]: It's consistent with the other options to omit the unit (_ms) here, but it probably would make it clearer. Not sure whether it's better to change it or not.

File: src/main.cpp [hunk 2]
```diff
@@ -1588,6 +1598,7 @@ int main(int argc, char* argv[])
   opt.reject_if_no_matching_ifcs = false;
   opt.dummy_app_server = "";
   opt.http_acr_logging = false;
+  opt.request_on_queue_timeout = 4000; // TODO decide on actual default
```
Comment (tomwatson1024) [inline src/main.cpp]: We agreed in the whiteboard that this needs some consideration once we've got a way to test overload - I'm just commenting so we remember!

File: src/thread_dispatcher.cpp [hunk 3]
```diff
@@ -90,6 +90,7 @@ static SNMP::EventAccumulatorByScopeTable* latency_table = NULL;
 static LoadMonitor* load_monitor = NULL;
 static SNMP::EventAccumulatorByScopeTable* queue_size_table = NULL;
 static ExceptionHandler* exception_handler = NULL;
+static unsigned long request_on_queue_timeout_us = 1;
```
Comment (tomwatson1024) [inline src/thread_dispatcher.cpp]: Why does this default to 1 over any other value? (more curiosity than anything else)

File: sprout-base.root/etc/init.d/sprout [hunk 4]
```diff
@@ -157,6 +157,7 @@ get_daemon_args()
         [ -z "$chronos_hostname" ] || chronos_hostname_arg="--chronos-hostname=$chronos_hostname"
         [ -z "$sprout_chronos_callback_uri" ] || sprout_chronos_callback_uri_arg="--sprout-chronos-callback-uri=$sprout_chronos_callback_uri"
         [ -z "$dummy_app_server" ] || dummy_app_server_arg="--dummy-app-server=$dummy_app_server"
+        [ -z "$request_on_queue_timeout" ] || request_on_queue_timeout_arg="--request-on-queue-timeout=$request_on_queue_timeout"
```
Comment (AlexHockey) [inline sprout-base.root/etc/init.d/sprout]: It's easier to add the new options to the block around line 286.

File: src/main.cpp [hunk 5]
```diff
@@ -1309,6 +1311,14 @@ static pj_status_t init_options(int argc, char* argv[], struct options* options)
       }
       break;
 
+    case OPT_REQUEST_ON_QUEUE_TIMEOUT:
+      {
+        VALIDATE_INT_PARAM(options->request_on_queue_timeout,
```
Comment (AlexHockey) [inline src/main.cpp]: This can't be zero or negative. Is there a better validation macro to use?

File: src/thread_dispatcher.cpp (hunk 3 above)
Comment (AlexHockey) [reply src/thread_dispatcher.cpp]: Does it need a default. Isn't the timeout a mandatory parameter to `init_thread_dispatcher`?

File: src/thread_dispatcher.cpp [hunk 6]
```diff
@@ -144,67 +145,112 @@ static int worker_thread(void* p)
...
-          {
-            TRC_ERROR("Exception Call-Id: %.*s (maybe)",
-                      ((pjsip_cid_hdr*)rdata->msg_info.cid)->id.slen,
-                      ((pjsip_cid_hdr*)rdata->msg_info.cid)->id.ptr);
-          }
-          if (rdata->msg_info.cseq != NULL)
+          TRC_ERROR("Failed to get timestamp: %s", strerror(errno));
+        }
+
+        if ((latency_us > (request_on_queue_timeout_us)) &&
```
Comment (AlexHockey) [inline src/thread_dispatcher.cpp]: Odd bracketing of `request_on_queue_timeout_us`

File: src/thread_dispatcher.cpp [hunk 7]
```diff
@@ -144,67 +145,112 @@ static int worker_thread(void* p)
...
+            SAS::Marker end_marker(trail, MARKER_ID_END, 2u);
+            SAS::report_marker(end_marker);
 
-          if ((rdata->msg_info.msg->type == PJSIP_REQUEST_MSG) &&
-             (rdata->msg_info.msg->line.req.method.id != PJSIP_ACK_METHOD))
-          {
-            TRC_DEBUG("Returning 500 response following exception");
             pjsip_retry_after_hdr* retry_after =
-                           pjsip_retry_after_hdr_create(rdata->tp_info.pool, 600);
+                           pjsip_retry_after_hdr_create(rdata->tp_info.pool, 0);
```
Comment (AlexHockey) [inline src/thread_dispatcher.cpp]: Why is 0 the right value to use for the Retry-After header? Regardless, this is probably worth a comment.

File: src/thread_dispatcher.cpp [hunk 8]
```diff
@@ -144,67 +145,112 @@ static int worker_thread(void* p)
...
+                                       NULL);
+            }
+
+            if (num_worker_threads == 1)
+            {
+              // There's only one worker thread, so we can't sensibly proceed.
+              exit(1);
+            }
+          }
+          CW_END
```
Comment (AlexHockey) [inline src/thread_dispatcher.cpp]: I've not the reviewed the catch block as it looked like it was just moved. Shout if this is wrong.

File: src/thread_dispatcher.cpp [hunk 9]
```diff
@@ -144,67 +145,112 @@ static int worker_thread(void* p)
...
-          load_monitor->request_complete(latency_us);
-        }
-        else
-        {
-          TRC_ERROR("Failed to get done timestamp: %s", strerror(errno));
+          if (me->stop_watch.read(latency_us))
+          {
+            TRC_DEBUG("Request latency = %ldus", latency_us);
+            latency_table->accumulate(latency_us);
+            load_monitor->request_complete(latency_us);
```
Comment (AlexHockey) [inline src/thread_dispatcher.cpp]: I think this means that the latency of requests that are rejected for having been on the queue for too long is included in this stat. I honestly don't know if that's right or not. Thoughts?

**Comment by tomwatson1024** [comment]
I'm going to go ahead and merge this into overload-integration, but any markups can be made on that branch.

**Pull request merged**